apps/server/src/modules/tool/common/service/common-tool-validation.service.ts

```diff
@@ -0,0 +1,81 @@
+import { Injectable } from '@nestjs/common';
+import { ValidationError } from '@shared/common';
+import {
+	ContextExternalToolDO,
+	CustomParameterDO,
+	CustomParameterEntryDO,
+	CustomParameterScope,
+	CustomParameterType,
+	ExternalToolDO,
+	SchoolExternalToolDO,
+} from '@shared/domain';
+import { isNaN } from 'lodash';
+
+const typeCheckers: { [key in CustomParameterType]: (val: string) => boolean } = {
+	[CustomParameterType.STRING]: () => true,
+	[CustomParameterType.NUMBER]: (val: string) => !isNaN(Number(val)),
+	[CustomParameterType.BOOLEAN]: (val: string) => val === 'true' || val === 'false',
+	[CustomParameterType.AUTO_CONTEXTID]: () => true,
+	[CustomParameterType.AUTO_CONTEXTNAME]: () => true,
+	[CustomParameterType.AUTO_SCHOOLID]: () => true,
+	[CustomParameterType.AUTO_SCHOOLNUMBER]: () => true,
+};
+
+export type ValidatableTool = SchoolExternalToolDO | ContextExternalToolDO;
+
+@Injectable()
+export class CommonToolValidationService {
+	public checkForDuplicateParameters(validatableTool: ValidatableTool): void {
+		const caseInsensitiveNames: string[] = validatableTool.parameters.map(({ name }: CustomParameterEntryDO) =>
+			name.toLowerCase()
+		);
+
+		const uniqueNames: Set<string> = new Set(caseInsensitiveNames);
+		if (uniqueNames.size !== validatableTool.parameters.length) {
+			throw new ValidationError(
+				`tool_param_duplicate: The tool ${validatableTool.id ?? ''} contains multiple of the same custom parameters.`
+			);
+		}
+	}
+
+	public checkCustomParameterEntries(loadedExternalTool: ExternalToolDO, validatableTool: ValidatableTool) {
+		if (loadedExternalTool.parameters) {
+			for (const param of loadedExternalTool.parameters) {
+				const foundEntry: CustomParameterEntryDO | undefined = validatableTool.parameters.find(
+					({ name }: CustomParameterEntryDO) => name.toLowerCase() === param.name.toLowerCase()
+				);
+				if (CustomParameterScope.SCHOOL === param.scope || CustomParameterScope.CONTEXT === param.scope) {
```

Comment:
This is new. All other things were moved. Only the tests were changed to consist the test structure